Fixed documentation and handles null topicPartition for KAFKA-12841

philipnee opened the pull request:
Jira: https://issues.apache.org/jira/browse/KAFKA-12841

Using the `InterceptorCallback` wrapper in the case of `ApiException` so that we will adhere correctly to the `Callback` contract for `onCompletion` specifying a valid (dummy) `TopicPartition`.   Removed some documentation from Callback.java that stated "except topicPartition" as we are assigning -1 to the topicPartition if it doesn't exist.  The changes is based on https://issues.apache.org/jira/browse/KAFKA-3303

### Committer Checklist (excluded from commit message)
 * [ ]  Verify design and implementation
 * [ ]  Verify test coverage and CI build status
 * [ ]  Verify documentation (including upgrade notes)

Commit:
KAFKA-12841: NPE from the provided metadata in client callback in case of ApiException

Using the InterceptorCallback wrapper in the case of ApiException so
that we will adhere correctly to the Callback contract for onCompletion
specifying a valid (dummy) TopicPartition.

Commit:
Whoops! Removing the debug stack trace printing from the test

philipnee committed:
topicPartition will also get -1 if not set, per KAFKA-3303

update documentation

philipnee committed:
topicPartition will also get -1 if not set, per KAFKA-3303

philipnee committed:
callback.java documentation

when metadata null, all fields are -1

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/producer/Callback.java:
```diff
@@ -25,10 +25,10 @@
      * A callback method the user can implement to provide asynchronous handling of request completion. This method will
      * be called when the record sent to the server has been acknowledged. When exception is not null in the callback,
-     * metadata will contain the special -1 value for all fields except for topicPartition, which will be valid.
+     * metadata will contain the special -1 value for all fields.
```
Please add a sentence specifying that we will return a partition ID of -1 if no partition could be chosen.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerInterceptors.java:
```diff
@@ -123,6 +122,10 @@ public void onSendError(ProducerRecord<K, V> record, TopicPartition interceptTop
         }
     }
 
+    public static <K, V> TopicPartition createTopicPartition(ProducerRecord<K, V> record) {
```
This is not a very good name since "creating a partition" usually means calling `createPartitions`. How about `extractTopicPartition` ?

cmccabe commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -1506,6 +1508,56 @@ public void testNullTopicName() {
+                try {
+                    assertNotNull(recordMetadata.topic());
+                } catch (NullPointerException e) {
+                    fail("Topic name should be valid even on send failure", e);
```
It's not necessary to do this. If you want to display a special error message when the assert fails, there is a three-argument form which lets you specify the error message.

cmccabe commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -1506,6 +1508,56 @@ public void testNullTopicName() {
+                assertEquals(invalidTopicName, recordMetadata.topic());
+
+                try {
+                    assertEquals(RecordMetadata.UNKNOWN_PARTITION, recordMetadata.partition());
```
As before, it's not necessary to do this. If you want to display a special error message when the assert fails, there is a three-argument form which lets you specify the error message.

philipnee committed:
Addressed PR comments

- specify topicPartition behavior
- method naming
- use 2 params assert to display proper error messages

philipnee committed:
Addressed PR comments

philipnee committed:
PR comment part 2

philipnee committed:
Merge branch 'pnee-12841' of github.com:philipnee/kafka into pnee-12841

cmccabe commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -1506,6 +1508,46 @@ public void testNullTopicName() {
+                assertNotNull(exception);
+                assertNotNull(recordMetadata);
+
+                assertNotNull(recordMetadata.topic(), "Topic name should be valid even on send failure");
```
can you please also check that the partition id gets set to -1

philipnee replied:
Done - see line 1545 in the test.

philipnee committed:
Update KafkaProducerTest.java

cmccabe commented:
test failures are not related.

cmccabe reviewed (approved):
LGTM

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -995,8 +995,17 @@ private Future<RecordMetadata> doSend(ProducerRecord<K, V> record, Callback call
+
+            // The onCompletion callback does expect a non-null metadata, but one will be created inside
+            // the interceptor's onCompletion implementation before the user's callback is invoked.
+            interceptCallback.onCompletion(null, e);
```
@philipnee : It seems that we may have introduced a bug here. interceptors.onAcknowledgement() is now called twice, once through interceptCallback.onCompletion() and another through interceptors.onSendError() in line 1010 below.

philipnee replied:
Hey @junrao - thanks for catching this.  Created a PR to address this bug: https://github.com/apache/kafka/pull/12064

junrao reviewed (commented):
@philipnee : Thanks for the PR. Added a comment below.
